relay/pkg/block/forwarder.go

```diff
@@ -0,0 +1,143 @@
+package block
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-log"
+	"github.com/keep-network/tbtc/relay/pkg/btc"
+	"github.com/keep-network/tbtc/relay/pkg/chain"
+)
+
+// TODO: Make the following refactoring:
+//  - rename `block` package to `header` (align all logging and stuff)
+//  - rename `Forwarder` to `Relay`
+//  - rename `RunForwarder` to `StartRelay`
+
+const (
+	// Size of the headers queue.
+	headersQueueSize = 50
+
+	// Maximum size of processed headers batch.
+	headersBatchSize = 5
+
+	// Maximum time for which the pulling process will wait for a single header
+	// to be delivered by the headers queue.
+	headerTimeout = 1 * time.Second
+
+	// Block duration of a Bitcoin difficulty epoch.
+	difficultyEpochDuration = 2016
+
+	// Duration for which the forwarder should rest after performing
+	// a push action.
+	forwarderPushingSleepTime = 45 * time.Second
+)
+
+var logger = log.Logger("relay-block-forwarder")
+
+// Forwarder takes blocks from the Bitcoin chain and forwards them to the
+// given host chain.
+type Forwarder struct {
+	btcChain  btc.Handle
+	hostChain chain.Handle
+
+	processedHeaders int
+
+	headersQueue chan *btc.Header
+	errChan      chan error
+
+	loopExitHandler func()
+}
+
+// RunForwarder creates an instance of the block forwarder and runs its
+// processing loops. The lifecycle of the forwarder can be managed using the
+// passed context.
+func RunForwarder(
+	ctx context.Context,
+	btcChain btc.Handle,
+	hostChain chain.Handle,
+) *Forwarder {
+	loopCtx, cancelLoopCtx := context.WithCancel(ctx)
+
+	forwarder := &Forwarder{
+		btcChain:        btcChain,
+		hostChain:       hostChain,
+		headersQueue:    make(chan *btc.Header, headersQueueSize),
+		errChan:         make(chan error, 1),
+		loopExitHandler: cancelLoopCtx,
+	}
+
+	go forwarder.pushingLoop(loopCtx)
+
+	return forwarder
+}
+
+func (f *Forwarder) pushingLoop(ctx context.Context) {
```

Comment:
Thoughts on renaming this function to `relayLoop`? (in a separate PR)

`pushingLoop` is confusing because we both pull and push inside this function.

Reply:
Nevermind, this is push to host chain, other comments clarified it.

Reply:
If we apply changes proposed in https://github.com/keep-network/tbtc/pull/767#discussion_r603954935 we will have `getHeadersFromQueue` and `pushHeadersToHostChain` so I think it won't be confusing anymore as pull/push naming will be only related with chains. Also `pushingLoop` fits better with the `pullingLoop` added in #769.